skimage/restoration/_inpaint.pyx

```diff
@@ -0,0 +1,57 @@
+import numpy as np
+import cython
+
+cimport numpy as cnp
+from .._shared.fused_numerics cimport np_floats
+
+
+@cython.boundscheck(False)  # Deactivate bounds checking
+@cython.wraparound(False)   # Deactivate negative indexing.
+@cython.cdivision(True)  # C style integer division
+cpdef Py_ssize_t _build_matrix_inner(
+    # starting index offsets
+    Py_ssize_t row_start,
+    Py_ssize_t known_start_idx,
+    Py_ssize_t unknown_start_idx,
+    # input arrays
+    const Py_ssize_t[::1] center_i,
+    const Py_ssize_t[::1] raveled_offsets,
+    np_floats[::1] coef_vals,
+    const char[::1] mask_flat,
+    np_floats[:, ::1] out_flat,
+    # output arrays
+    Py_ssize_t[::1] row_idx_known,
+    np_floats[:, ::1] data_known,
+    Py_ssize_t[::1] row_idx_unknown,
+    Py_ssize_t[::1] col_idx_unknown,
+    np_floats[::1] data_unknown
+):
+    """Fill values in *_known and *_unkown"""
+    cdef:
+        Py_ssize_t ch
+        Py_ssize_t row_idx, known_idx, unknown_idx, loc, n_known
+        np_floats cval
+        int nchannels = data_known.shape[1]
+
+    row_idx = row_start
+    known_idx = known_start_idx
+    unknown_idx = unknown_start_idx
+
+    for c_i in center_i:
+        n_known = 0
+        for offset, cval in zip(raveled_offsets, coef_vals):
+            loc = c_i + offset
```

Comment:
This ends up being less efficient than the prior code, probably due to use of zip so I have reverted part of the Cython changes.

I did keep the mask as `char` change, though!